// @credo-ts/anoncreds/build/utils/indyIdentifiers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getQualifiedDidIndyRevocationRegistryDefinition = exports.isQualifiedRevocationRegistryDefinition = exports.getUnqualifiedDidIndyRevocationRegistryDefinition = exports.isUnqualifiedDidIndyRevocationRegistryDefinition = exports.getQualifiedDidIndyCredentialDefinition = exports.isQualifiedDidIndyCredentialDefinition = exports.getUnqualifiedDidIndyCredentialDefinition = exports.isUnqualifiedDidIndyCredentialDefinition = exports.getQualifiedDidIndySchema = exports.isQualifiedDidIndySchema = exports.getUnqualifiedDidIndySchema = exports.isUnqualifiedDidIndySchema = exports.getQualifiedDidIndyDid = exports.isIndyDid = exports.getUnQualifiedDidIndyDid = exports.getIndyNamespaceFromIndyDid = exports.parseIndyRevocationRegistryId = exports.parseIndyCredentialDefinitionId = exports.parseIndySchemaId = exports.parseIndyDid = exports.isDidIndyRevocationRegistryId = exports.isDidIndyCredentialDefinitionId = exports.isDidIndySchemaId = exports.isUnqualifiedSchemaId = exports.isUnqualifiedRevocationRegistryId = exports.isUnqualifiedCredentialDefinitionId = exports.isUnqualifiedIndyDid = exports.getUnqualifiedRevocationRegistryDefinitionId = exports.getUnqualifiedCredentialDefinitionId = exports.getUnqualifiedSchemaId = exports.didIndyRegex = exports.didIndyRevocationRegistryIdRegex = exports.unqualifiedRevocationRegistryIdRegex = exports.didIndyCredentialDefinitionIdRegex = exports.unqualifiedCredentialDefinitionIdRegex = exports.unqualifiedIndyDidRegex = exports.unqualifiedSchemaVersionRegex = exports.didIndySchemaIdRegex = exports.unqualifiedSchemaIdRegex = void 0;
const core_1 = require("@credo-ts/core");
const didIndyAnonCredsBase = /(did:indy:((?:[a-z][_a-z0-9-]*)(?::[a-z][_a-z0-9-]*)?):([1-9A-HJ-NP-Za-km-z]{21,22}))\/anoncreds\/v0/;
// <namespaceIdentifier>:2:<schemaName>:<schemaVersion>
exports.unqualifiedSchemaIdRegex = /^([a-zA-Z0-9]{21,22}):2:(.+):([0-9.]+)$/;
// did:indy:<namespace>:<namespaceIdentifier>/anoncreds/v0/SCHEMA/<schemaName>/<schemaVersion>
exports.didIndySchemaIdRegex = new RegExp(`^${didIndyAnonCredsBase.source}/SCHEMA/(.+)/([0-9.]+)$`);
exports.unqualifiedSchemaVersionRegex = /^(\d+\.)?(\d+\.)?(\*|\d+)$/;
exports.unqualifiedIndyDidRegex = /^(did:sov:)?[a-zA-Z0-9]{21,22}$/;
// <namespaceIdentifier>:3:CL:<schemaSeqNo>:<tag>
exports.unqualifiedCredentialDefinitionIdRegex = /^([a-zA-Z0-9]{21,22}):3:CL:([1-9][0-9]*):(.+)$/;
// did:indy:<namespace>:<namespaceIdentifier>/anoncreds/v0/CLAIM_DEF/<schemaSeqNo>/<tag>
exports.didIndyCredentialDefinitionIdRegex = new RegExp(`^${didIndyAnonCredsBase.source}/CLAIM_DEF/([1-9][0-9]*)/(.+)$`);
// <namespaceIdentifier>:4:<namespaceIdentifier>:3:CL:<schemaSeqNo>:<credentialDefinitionTag>:CL_ACCUM:<revocationRegistryTag>
exports.unqualifiedRevocationRegistryIdRegex = /^([a-zA-Z0-9]{21,22}):4:[a-zA-Z0-9]{21,22}:3:CL:([1-9][0-9]*):(.+):CL_ACCUM:(.+)$/;
// did:indy:<namespace>:<namespaceIdentifier>/anoncreds/v0/REV_REG_DEF/<schemaSeqNo>/<credentialDefinitionTag>/<revocationRegistryTag>
exports.didIndyRevocationRegistryIdRegex = new RegExp(`^${didIndyAnonCredsBase.source}/REV_REG_DEF/([1-9][0-9]*)/(.+)/(.+)$`);
exports.didIndyRegex = /^did:indy:((?:[a-z][_a-z0-9-]*)(?::[a-z][_a-z0-9-]*)?):([1-9A-HJ-NP-Za-km-z]{21,22})$/;
function getUnqualifiedSchemaId(unqualifiedDid, name, version) {
    return `${unqualifiedDid}:2:${name}:${version}`;
}
exports.getUnqualifiedSchemaId = getUnqualifiedSchemaId;
function getUnqualifiedCredentialDefinitionId(unqualifiedDid, schemaSeqNo, tag) {
    return `${unqualifiedDid}:3:CL:${schemaSeqNo}:${tag}`;
}
exports.getUnqualifiedCredentialDefinitionId = getUnqualifiedCredentialDefinitionId;
// TZQuLp43UcYTdtc3HewcDz:4:TZQuLp43UcYTdtc3HewcDz:3:CL:98158:BaustellenzertifikateNU1:CL_ACCUM:1-100
function getUnqualifiedRevocationRegistryDefinitionId(unqualifiedDid, schemaSeqNo, credentialDefinitionTag, revocationRegistryTag) {
    return `${unqualifiedDid}:4:${unqualifiedDid}:3:CL:${schemaSeqNo}:${credentialDefinitionTag}:CL_ACCUM:${revocationRegistryTag}`;
}
exports.getUnqualifiedRevocationRegistryDefinitionId = getUnqualifiedRevocationRegistryDefinitionId;
function isUnqualifiedIndyDid(did) {
    return exports.unqualifiedIndyDidRegex.test(did);
}
exports.isUnqualifiedIndyDid = isUnqualifiedIndyDid;
function isUnqualifiedCredentialDefinitionId(credentialDefinitionId) {
    return exports.unqualifiedCredentialDefinitionIdRegex.test(credentialDefinitionId);
}
exports.isUnqualifiedCredentialDefinitionId = isUnqualifiedCredentialDefinitionId;
function isUnqualifiedRevocationRegistryId(revocationRegistryId) {
    return exports.unqualifiedRevocationRegistryIdRegex.test(revocationRegistryId);
}
exports.isUnqualifiedRevocationRegistryId = isUnqualifiedRevocationRegistryId;
function isUnqualifiedSchemaId(schemaId) {
    return exports.unqualifiedSchemaIdRegex.test(schemaId);
}
exports.isUnqualifiedSchemaId = isUnqualifiedSchemaId;
function isDidIndySchemaId(schemaId) {
    return exports.didIndySchemaIdRegex.test(schemaId);
}
exports.isDidIndySchemaId = isDidIndySchemaId;
function isDidIndyCredentialDefinitionId(credentialDefinitionId) {
    return exports.didIndyCredentialDefinitionIdRegex.test(credentialDefinitionId);
}
exports.isDidIndyCredentialDefinitionId = isDidIndyCredentialDefinitionId;
function isDidIndyRevocationRegistryId(revocationRegistryId) {
    return exports.didIndyRevocationRegistryIdRegex.test(revocationRegistryId);
}
exports.isDidIndyRevocationRegistryId = isDidIndyRevocationRegistryId;
function parseIndyDid(did) {
    const match = did.match(exports.didIndyRegex);
    if (match) {
        const [, namespace, namespaceIdentifier] = match;
        return { namespace, namespaceIdentifier };
    }
    else {
        throw new core_1.CredoError(`${did} is not a valid did:indy did`);
    }
}
exports.parseIndyDid = parseIndyDid;
function parseIndySchemaId(schemaId) {
    const didIndyMatch = schemaId.match(exports.didIndySchemaIdRegex);
    if (didIndyMatch) {
        const [, did, namespace, namespaceIdentifier, schemaName, schemaVersion] = didIndyMatch;
        return {
            did,
            namespaceIdentifier,
            schemaName,
            schemaVersion,
            namespace,
        };
    }
    const legacyMatch = schemaId.match(exports.unqualifiedSchemaIdRegex);
    if (legacyMatch) {
        const [, did, schemaName, schemaVersion] = legacyMatch;
        return {
            did,
            namespaceIdentifier: did,
            schemaName,
            schemaVersion,
        };
    }
    throw new Error(`Invalid schema id: ${schemaId}`);
}
exports.parseIndySchemaId = parseIndySchemaId;
function parseIndyCredentialDefinitionId(credentialDefinitionId) {
    const didIndyMatch = credentialDefinitionId.match(exports.didIndyCredentialDefinitionIdRegex);
    if (didIndyMatch) {
        const [, did, namespace, namespaceIdentifier, schemaSeqNo, tag] = didIndyMatch;
        return {
            did,
            namespaceIdentifier,
            schemaSeqNo,
            tag,
            namespace,
        };
    }
    const legacyMatch = credentialDefinitionId.match(exports.unqualifiedCredentialDefinitionIdRegex);
    if (legacyMatch) {
        const [, did, schemaSeqNo, tag] = legacyMatch;
        return {
            did,
            namespaceIdentifier: did,
            schemaSeqNo,
            tag,
        };
    }
    throw new Error(`Invalid credential definition id: ${credentialDefinitionId}`);
}
exports.parseIndyCredentialDefinitionId = parseIndyCredentialDefinitionId;
function parseIndyRevocationRegistryId(revocationRegistryId) {
    const didIndyMatch = revocationRegistryId.match(exports.didIndyRevocationRegistryIdRegex);
    if (didIndyMatch) {
        const [, did, namespace, namespaceIdentifier, schemaSeqNo, credentialDefinitionTag, revocationRegistryTag] = didIndyMatch;
        return {
            did,
            namespaceIdentifier,
            schemaSeqNo,
            credentialDefinitionTag,
            revocationRegistryTag,
            namespace,
        };
    }
    const legacyMatch = revocationRegistryId.match(exports.unqualifiedRevocationRegistryIdRegex);
    if (legacyMatch) {
        const [, did, schemaSeqNo, credentialDefinitionTag, revocationRegistryTag] = legacyMatch;
        return {
            did,
            namespaceIdentifier: did,
            schemaSeqNo,
            credentialDefinitionTag,
            revocationRegistryTag,
        };
    }
    throw new Error(`Invalid revocation registry id: ${revocationRegistryId}`);
}
exports.parseIndyRevocationRegistryId = parseIndyRevocationRegistryId;
function getIndyNamespaceFromIndyDid(identifier) {
    let namespace;
    if (isDidIndySchemaId(identifier)) {
        namespace = parseIndySchemaId(identifier).namespace;
    }
    else if (isDidIndyCredentialDefinitionId(identifier)) {
        namespace = parseIndyCredentialDefinitionId(identifier).namespace;
    }
    else if (isDidIndyRevocationRegistryId(identifier)) {
        namespace = parseIndyRevocationRegistryId(identifier).namespace;
    }
    else {
        namespace = parseIndyDid(identifier).namespace;
    }
    if (!namespace)
        throw new core_1.CredoError(`Cannot get indy namespace of identifier '${identifier}'`);
    return namespace;
}
exports.getIndyNamespaceFromIndyDid = getIndyNamespaceFromIndyDid;
function getUnQualifiedDidIndyDid(identifier) {
    if (isDidIndySchemaId(identifier)) {
        const { schemaName, schemaVersion, namespaceIdentifier } = parseIndySchemaId(identifier);
        return getUnqualifiedSchemaId(namespaceIdentifier, schemaName, schemaVersion);
    }
    else if (isDidIndyCredentialDefinitionId(identifier)) {
        const { schemaSeqNo, tag, namespaceIdentifier } = parseIndyCredentialDefinitionId(identifier);
        return getUnqualifiedCredentialDefinitionId(namespaceIdentifier, schemaSeqNo, tag);
    }
    else if (isDidIndyRevocationRegistryId(identifier)) {
        const { namespaceIdentifier, schemaSeqNo, credentialDefinitionTag, revocationRegistryTag } = parseIndyRevocationRegistryId(identifier);
        return getUnqualifiedRevocationRegistryDefinitionId(namespaceIdentifier, schemaSeqNo, credentialDefinitionTag, revocationRegistryTag);
    }
    const { namespaceIdentifier } = parseIndyDid(identifier);
    return namespaceIdentifier;
}
exports.getUnQualifiedDidIndyDid = getUnQualifiedDidIndyDid;
function isIndyDid(identifier) {
    return identifier.startsWith('did:indy:');
}
exports.isIndyDid = isIndyDid;
function getQualifiedDidIndyDid(identifier, namespace) {
    if (isIndyDid(identifier))
        return identifier;
    if (!namespace || typeof namespace !== 'string') {
        throw new core_1.CredoError('Missing required indy namespace');
    }
    if (isUnqualifiedSchemaId(identifier)) {
        const { namespaceIdentifier, schemaName, schemaVersion } = parseIndySchemaId(identifier);
        const schemaId = `did:indy:${namespace}:${namespaceIdentifier}/anoncreds/v0/SCHEMA/${schemaName}/${schemaVersion}`;
        return schemaId;
    }
    else if (isUnqualifiedCredentialDefinitionId(identifier)) {
        const { namespaceIdentifier, schemaSeqNo, tag } = parseIndyCredentialDefinitionId(identifier);
        const credentialDefinitionId = `did:indy:${namespace}:${namespaceIdentifier}/anoncreds/v0/CLAIM_DEF/${schemaSeqNo}/${tag}`;
        return credentialDefinitionId;
    }
    else if (isUnqualifiedRevocationRegistryId(identifier)) {
        const { namespaceIdentifier, schemaSeqNo, credentialDefinitionTag, revocationRegistryTag } = parseIndyRevocationRegistryId(identifier);
        const revocationRegistryId = `did:indy:${namespace}:${namespaceIdentifier}/anoncreds/v0/REV_REG_DEF/${schemaSeqNo}/${credentialDefinitionTag}/${revocationRegistryTag}`;
        return revocationRegistryId;
    }
    else if (isUnqualifiedIndyDid(identifier)) {
        return `did:indy:${namespace}:${identifier}`;
    }
    else {
        throw new core_1.CredoError(`Cannot created qualified indy identifier for '${identifier}' with namespace '${namespace}'`);
    }
}
exports.getQualifiedDidIndyDid = getQualifiedDidIndyDid;
// -- schema -- //
function isUnqualifiedDidIndySchema(schema) {
    return isUnqualifiedIndyDid(schema.issuerId);
}
exports.isUnqualifiedDidIndySchema = isUnqualifiedDidIndySchema;
function getUnqualifiedDidIndySchema(schema) {
    if (isUnqualifiedDidIndySchema(schema))
        return Object.assign({}, schema);
    if (!isIndyDid(schema.issuerId)) {
        throw new core_1.CredoError(`IssuerId '${schema.issuerId}' is not a valid qualified did-indy did.`);
    }
    const issuerId = getUnQualifiedDidIndyDid(schema.issuerId);
    return Object.assign(Object.assign({}, schema), { issuerId });
}
exports.getUnqualifiedDidIndySchema = getUnqualifiedDidIndySchema;
function isQualifiedDidIndySchema(schema) {
    return !isUnqualifiedIndyDid(schema.issuerId);
}
exports.isQualifiedDidIndySchema = isQualifiedDidIndySchema;
function getQualifiedDidIndySchema(schema, namespace) {
    if (isQualifiedDidIndySchema(schema))
        return Object.assign({}, schema);
    return Object.assign(Object.assign({}, schema), { issuerId: getQualifiedDidIndyDid(schema.issuerId, namespace) });
}
exports.getQualifiedDidIndySchema = getQualifiedDidIndySchema;
// -- credential definition -- //
function isUnqualifiedDidIndyCredentialDefinition(anonCredsCredentialDefinition) {
    return (isUnqualifiedIndyDid(anonCredsCredentialDefinition.issuerId) &&
        isUnqualifiedSchemaId(anonCredsCredentialDefinition.schemaId));
}
exports.isUnqualifiedDidIndyCredentialDefinition = isUnqualifiedDidIndyCredentialDefinition;
function getUnqualifiedDidIndyCredentialDefinition(anonCredsCredentialDefinition) {
    if (isUnqualifiedDidIndyCredentialDefinition(anonCredsCredentialDefinition)) {
        return Object.assign({}, anonCredsCredentialDefinition);
    }
    const issuerId = getUnQualifiedDidIndyDid(anonCredsCredentialDefinition.issuerId);
    const schemaId = getUnQualifiedDidIndyDid(anonCredsCredentialDefinition.schemaId);
    return Object.assign(Object.assign({}, anonCredsCredentialDefinition), { issuerId, schemaId });
}
exports.getUnqualifiedDidIndyCredentialDefinition = getUnqualifiedDidIndyCredentialDefinition;
function isQualifiedDidIndyCredentialDefinition(anonCredsCredentialDefinition) {
    return (!isUnqualifiedIndyDid(anonCredsCredentialDefinition.issuerId) &&
        !isUnqualifiedSchemaId(anonCredsCredentialDefinition.schemaId));
}
exports.isQualifiedDidIndyCredentialDefinition = isQualifiedDidIndyCredentialDefinition;
function getQualifiedDidIndyCredentialDefinition(anonCredsCredentialDefinition, namespace) {
    if (isQualifiedDidIndyCredentialDefinition(anonCredsCredentialDefinition))
        return Object.assign({}, anonCredsCredentialDefinition);
    return Object.assign(Object.assign({}, anonCredsCredentialDefinition), { issuerId: getQualifiedDidIndyDid(anonCredsCredentialDefinition.issuerId, namespace), schemaId: getQualifiedDidIndyDid(anonCredsCredentialDefinition.schemaId, namespace) });
}
exports.getQualifiedDidIndyCredentialDefinition = getQualifiedDidIndyCredentialDefinition;
// -- revocation registry definition -- //
function isUnqualifiedDidIndyRevocationRegistryDefinition(revocationRegistryDefinition) {
    return (isUnqualifiedIndyDid(revocationRegistryDefinition.issuerId) &&
        isUnqualifiedCredentialDefinitionId(revocationRegistryDefinition.credDefId));
}
exports.isUnqualifiedDidIndyRevocationRegistryDefinition = isUnqualifiedDidIndyRevocationRegistryDefinition;
function getUnqualifiedDidIndyRevocationRegistryDefinition(revocationRegistryDefinition) {
    if (isUnqualifiedDidIndyRevocationRegistryDefinition(revocationRegistryDefinition)) {
        return Object.assign({}, revocationRegistryDefinition);
    }
    const issuerId = getUnQualifiedDidIndyDid(revocationRegistryDefinition.issuerId);
    const credDefId = getUnQualifiedDidIndyDid(revocationRegistryDefinition.credDefId);
    return Object.assign(Object.assign({}, revocationRegistryDefinition), { issuerId, credDefId });
}
exports.getUnqualifiedDidIndyRevocationRegistryDefinition = getUnqualifiedDidIndyRevocationRegistryDefinition;
function isQualifiedRevocationRegistryDefinition(revocationRegistryDefinition) {
    return (!isUnqualifiedIndyDid(revocationRegistryDefinition.issuerId) &&
        !isUnqualifiedCredentialDefinitionId(revocationRegistryDefinition.credDefId));
}
exports.isQualifiedRevocationRegistryDefinition = isQualifiedRevocationRegistryDefinition;
function getQualifiedDidIndyRevocationRegistryDefinition(revocationRegistryDefinition, namespace) {
    if (isQualifiedRevocationRegistryDefinition(revocationRegistryDefinition))
        return Object.assign({}, revocationRegistryDefinition);
    return Object.assign(Object.assign({}, revocationRegistryDefinition), { issuerId: getQualifiedDidIndyDid(revocationRegistryDefinition.issuerId, namespace), credDefId: getQualifiedDidIndyDid(revocationRegistryDefinition.credDefId, namespace) });
}
exports.getQualifiedDidIndyRevocationRegistryDefinition = getQualifiedDidIndyRevocationRegistryDefinition;
