// @credo-ts/anoncreds/build/repository/AnonCredsRevocationRegistryDefinitionRecord.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnonCredsRevocationRegistryDefinitionRecord = void 0;
const core_1 = require("@credo-ts/core");
class AnonCredsRevocationRegistryDefinitionRecord extends core_1.BaseRecord {
    constructor(props) {
        var _a;
        super();
        this.type = AnonCredsRevocationRegistryDefinitionRecord.type;
        if (props) {
            this.id = (_a = props.id) !== null && _a !== void 0 ? _a : core_1.utils.uuid();
            this.revocationRegistryDefinitionId = props.revocationRegistryDefinitionId;
            this.revocationRegistryDefinition = props.revocationRegistryDefinition;
        }
    }
    getTags() {
        return Object.assign(Object.assign({}, this._tags), { revocationRegistryDefinitionId: this.revocationRegistryDefinitionId, credentialDefinitionId: this.revocationRegistryDefinition.credDefId });
    }
}
exports.AnonCredsRevocationRegistryDefinitionRecord = AnonCredsRevocationRegistryDefinitionRecord;
AnonCredsRevocationRegistryDefinitionRecord.type = 'AnonCredsRevocationRegistryDefinitionRecord';
